packages/language-server/src/plugins/typescript/features/CodeActionsProvider.ts

```diff
@@ -264,324 +542,504 @@ export class CodeActionsProviderImpl implements CodeActionsProvider {
+        identifier: TextDocumentIdentifier,
+        fileName: string,
+        lang: ts.LanguageService
+    ) {
+        const checkedFixIds = new Set<FixId>();
+        const fixAll: CodeAction[] = [];
+
+        for (const codeFix of codeFixes) {
+            if (!codeFix.fixId || !codeFix.fixAllDescription || checkedFixIds.has(codeFix.fixId)) {
+                continue;
+            }
+
+            // we have custom fix for import
+            // check it again if fix-all might be necessary
+            if (codeFix.fixName === FIX_IMPORT_FIX_NAME) {
+                const allCannotFindNameDiagnostics = lang
+                    .getSemanticDiagnostics(fileName)
+                    .filter((diagnostic) => diagnostic.code === DiagnosticCode.CANNOT_FIND_NAME);
+
+                if (allCannotFindNameDiagnostics.length < 2) {
+                    checkedFixIds.add(codeFix.fixId);
+                    continue;
+                }
+            }
+
+            const codeAction = CodeAction.create(
+                codeFix.fixAllDescription,
+                CodeActionKind.QuickFix
+            );
+
+            const data: QuickFixAllResolveInfo = {
+                ...identifier,
+                fixName: codeFix.fixName,
+                fixId: codeFix.fixId
+            };
+
+            codeAction.data = data;
+            checkedFixIds.add(codeFix.fixId);
+            fixAll.push(codeAction);
+        }
+
+        return fixAll;
+    }
+
     /**
      * import quick fix requires the symbol name to be the same as where it's defined.
      * But we have suffix on component default export to prevent conflict with
      * a local variable. So we use auto-import completion as a workaround here.
      */
     private getComponentImportQuickFix(
-        start: number,
-        end: number,
+        document: Document,
         lang: ts.LanguageService,
         tsDoc: DocumentSnapshot,
         userPreferences: ts.UserPreferences,
         diagnostics: Diagnostic[],
         formatCodeSetting: ts.FormatCodeSettings
-    ): readonly ts.CodeFixAction[] | undefined {
+    ): CustomFixCannotFindNameInfo[] | undefined {
         const sourceFile = lang.getProgram()?.getSourceFile(tsDoc.filePath);
 
         if (!sourceFile) {
             return;
         }
 
-        const node = findContainingNode(
-            sourceFile,
-            {
-                start,
-                length: end - start
-            },
-            (node): node is ts.JsxOpeningLikeElement | ts.JsxClosingElement | ts.Identifier =>
-                ts.isCallExpression(node.parent) &&
-                ts.isIdentifier(node.parent.expression) &&
-                node.parent.expression.text === '__sveltets_2_ensureComponent' &&
-                ts.isIdentifier(node)
-        );
+        const nameToPosition = new Map<string, number>();
 
-        if (!node) {
-            return;
-        }
+        for (const diagnostic of diagnostics) {
```

Comment:
I didn't revert this part when switching to the virtual file method because you can select a range to start code action. So there are places where it will have multiple positions.